daemon/compress.c

```diff
@@ -0,0 +1,75 @@
+
+#include <str.h>
+#include <zlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include "log.h"
+#include "udp_listener.h"
+#include "str.h"
+#include "compress.h"
+
+#define windowBits 15
+
+#define GZIP_ENCODING 16
+#define ENABLE_ZLIB_GZIP 32
+
+static void init_stream(z_stream *stream);
+
+static void init_stream(z_stream *stream)
+{
+	stream->zalloc = Z_NULL;
+	stream->zfree = Z_NULL;
+	stream->opaque = Z_NULL;
+}
+
+
+// compress_data compress the data of the given length.
+// returns compressed buffer length and replace the original buffer.
+int compress_data(char *buf, int len) {
+	uLong comp_len = compressBound(len);
+    char tmp[comp_len];
+    int ret;
+
+    ret = compress2((Bytef *)tmp, &comp_len, (Bytef *)buf, len, Z_BEST_COMPRESSION);
+    if (ret != Z_OK) {
+        return -1;
+    }
+
+    memcpy(buf, tmp, comp_len);
+    return comp_len;
+}
+
+// uncompress_data uncompress the compressed data with given length.
+// it compare the first byte to check the compressed data or not.
+int uncompress_data(char *buf, int len, char *buf_tmp, int buf_size) {
```

Comment:
Same with the temporary buffer here, I'd just use different in and out buffers.

Reply:
Yes, that might be better.
But if you see where this function is used, you can understand it.

This uncompress_Data is being used in the infinite loop at the udp_listener_incoming(). So, rather than use the different buffer and make them allocating/deallocating the memory all the time for each message receiving, add the args for buffer and reuse them all the time is better.

Reply:
I still have the same idea, but I'm OK to fix it if you want to. :)